### ISSUE TITLE
Add Pseudo Mods for Crit and Spell/Ele/etc. Damage

### ISSUE THREAD

**Issue by trackpete** [trackpete/exiletools-indexer#38]
Suggested here: https://www.reddit.com/r/pathofexile/comments/3tz0y8/discussion_how_would_you_improve_poetrade/cxagsj2?context=3

This type of pseudo mod would be a great back-end addition to the index. For example:

```
              "#% increased Critical Strike Multiplier" : {
              "#% increased Critical Strike Multiplier for Spells" : {
              "#% increased Critical Strike Multiplier while Dual Wielding" : {
              "#% increased Critical Strike Multiplier with Cold Skills" : {
              "#% increased Critical Strike Multiplier with Elemental Skills" : {
              "#% increased Critical Strike Multiplier with Fire Skills" : {
              "#% increased Critical Strike Multiplier with Lightning Skills" : {
              "#% increased Critical Strike Multiplier with One Handed Melee Weapons" : {
              "#% increased Critical Strike Multiplier with Two Handed Melee Weapons" : {
              "#% increased Global Critical Strike Multiplier" : {
              "#% increased Melee Critical Strike Multiplier" : {
              "#% reduced Global Critical Strike Multiplier" : {
```

These mods could be broken down into:

- modsPseudo.critMulti.spells.all
- modsPseudo.critMulti.spells.elemental
- modsPseudo.critMulti.spells.cold
- modsPseudo.critMulti.spells.fire ... etc.

Mods that can be combined into pseudo (beyond just the current +total) should include global mods that have conditions:

- Critical Strike Chance (Global)
- Critical Strike Multiplier (Global)
- Increased Damage (Global)
- Spell/Melee/Ele/Chaos Damage (Global)
- Attack Speed?
- DOT?
- Chance to Block?
- Gain X% Damage as ?
- Burns?
- Minion damage / speed?

This will take a bit of research to break these down and categorize them all, but it should be a valuable addition to the extracted data about an item. Jewels will be a good place to start, for example:

```
+10% attack speed
+10% attack speed while dual wielding
+10% attack speed with daggers
```

Should surface with pseudomods like:

```
modsPseudo.Attack Speed #%.Global: 10
modsPseudo.Attack Speed #%.Daggers: 20
modsPseudo.Attack Speed #%.Dual Wield: 20
modsPseudo.Attack Speed #%.Dual Wield Daggers: 30
```

Some of these can be done programmatically while others (+fire = +ele) will need to be done with lookup trees.

**Comment by trackpete** [comment]
Starting this with `Attack Speed`, some thoughts:

We should ignore weapons, because those are weird. Unless we want to create a pseudo mod for claws that is `increased attack speed when on low life with movement skills` ?? lol no.

Jewels are a good place to start. I don't see anything else that gives multiple attack speed bonuses to be created via pseudo stats.

```
mods.Jewel.explicit.#% increased Attack Speed
mods.Jewel.explicit.#% increased Attack Speed while Dual Wielding
mods.Jewel.explicit.#% increased Attack Speed while holding a Shield
mods.Jewel.explicit.#% increased Attack Speed with Axes
mods.Jewel.explicit.#% increased Attack Speed with Bows
mods.Jewel.explicit.#% increased Attack Speed with Claws
mods.Jewel.explicit.#% increased Attack Speed with Daggers
mods.Jewel.explicit.#% increased Attack Speed with Maces
mods.Jewel.explicit.#% increased Attack Speed with One Handed Melee Weapons
mods.Jewel.explicit.#% increased Attack Speed with Staves
mods.Jewel.explicit.#% increased Attack Speed with Swords
mods.Jewel.explicit.#% increased Attack Speed with Two Handed Melee Weapons
mods.Jewel.explicit.#% increased Attack Speed with Wands
```

We don't necessarily need to limit these to logical pairings, we can let the player sort them out. So, take the following Jewel:

```
8fcb98e6b673cfc81c751dd710f4052d

   "mods" : {
      "Jewel" : {
         "explicit" : {
            "#% increased Attack Speed" : 5,
            "#% increased Attack Speed with Two Handed Melee Weapons" : 6,
            "#% increased Physical Damage with Swords" : 15
         }
      }
   },
```

Add the following:

```
   "modsPseudo" : {
      "#% Total increased Attack Speed with Two Handed Melee Weapons" : 11
   },
```

**Comment by trackpete** [comment]
Okay, a proof of concept is there but it's very basic - too basic. It will only add base attack speed to one modifier. The problem is handling complex items like this imaginary gem:

```
   "modsTotal" : {
      "#% increased Attack Speed" : 4,
      "#% increased Attack Speed with Swords" : 8,
      "#% increased Attack Speed with One Handed Melee Weapons" : 6,
      "#% increased Attack Speed with Two Handed Melee Weapons" : 8,
      "#% increased Attack Speed while Dual Wielding" : 6,
      "#% increased Attack Speed while holding a Shield" : 5
   }
```

The pseudo mod output needs to take these all into account to output something like:

```
#% Total increased Attack Speed with One Handed Melee Weapons : 10
#% Total increased Attack Speed with One Handed Melee Weapons while holding a Shield: 15
#% Total increased Attack Speed with Swords: 18
#% Total increased Attack Speed with Swords while holding a Shield: 23
#% Total increased Attack Speed while Dual Wielding: 10
#% Total increased Attack Speed while Dual Wielding with One Handed Melee Weapons: 16
#% Total increased Attack Speed while Dual Wielding with Swords: 24
#% Total increased Attack Speed with Two Handed Melee Weapons: 12
#% Total increased Attack Speed with Two Handed Swords: 20

```

It needs to know One Handed Melee Weapons can be Dual Wielded or used while holding a Shield, it needs to know Swords can be Dual Wielded, and thus also needs to know that Dual Wielding Swords gives you the One Handed Bonus. It has to know Swords can also be Two Handed, and that Two Handed Swords can't be Dual Wielded or used while holding a shield.

I think this will require a lookup table, at the very least to say something like:

* One Handed Melee Weapons include X, Y, Z
* Two Handed Melee Weapons include A, B, C
* Dual Wield can include any One Handed Melee Weapon
* Shield can include any One Handed Melee Weapon
* generic attack speed increases apply to everything

We can't do this by straight programmatic addition, or we'd end up with contradictory and bizarre stuff like:

```
#% Total increased Attack Speed while Dual Wielding with Swords with One Handed Melee Weapons while holding a Shield with Two Handed Melee Weapons: 999,999,999 *bleep bloop*
```

I've pulled the subroutine for now and will have to revisit this later. On the plus side, it looks like whatever I figure out for this will work for at least damage bonuses and attack speed on weapons, then I need to figure out what to do about all the other possible pseudo mods.

**Comment by trackpete** [comment]
I've decided that since I can't proceed doing this with a programmatic evaluation, the best way forward will be to attempt to crowdsource this data from reddit/etc. and hope people will contribute. I will make a new case to track this.

**Comment by thirdy** [comment]
Wonder if this can be handled via ES Query.

Google'd "elastic search query sum of fields", and I saw this:

http://stackoverflow.com/questions/24444966/elastic-search-find-sum-of-two-fields-in-single-query

Sounds like it is possible. Can't try it now though, I'm at work.


On Mon, Dec 21, 2015 at 10:38 PM, Pete Waterman <notifications@github.com>
wrote:

> Closed #38 <https://github.com/trackpete/exiletools-indexer/issues/38>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/trackpete/exiletools-indexer/issues/38#event-497481012>
> .
>



-- 
--
http://sites.google.com/site/thirdyderivera/
